enhancements/network/network-policy-audit-logging.md

```diff
@@ -0,0 +1,357 @@
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantA
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "notice" }'
+```
+
+or
+
+```go
+kind: Namespace
+apiVersion: v1
+metadata:
+  name: tenantB
+  annotations:
+    k8s.ovn.org/acl-logging: '{ "deny": "notice" }'
+```
+
+The logging can be activated for either allow, drop, or allow and drop actions. The severity must be one of `alert`, `warning`,
+`notice`, `info`, or `debug` as described in [OVN documentation](http://www.openvswitch.org/support/dist-docs/ovn-nbctl.8.html).
+
+each level is listed in order or descending severity
+
+```go
+alert    A major failure forced a process to abort.
+
+warning  A high-level operation or a subsystem failed.  Attention
+         is warranted.
+
+notice   A low-level operation failed, but higher-level subsystems
+         may be able to recover.
+
+info     Information that may be useful in retrospect when
+         investigating a problem.
+
+debug    Information useful only to someone with intricate
+         knowledge of the system, or that would commonly cause too-
+         voluminous log output.  Log messages at this level are not
+         logged by default.
+```
+
+### User Stories
+
+#### Story 1
+
+As an Openshift user I want to see what Network Policies and accompanied ACLs are dropping the most traffic in a cluster. For example, when a packet runs into a `drop` ACL it will log a message similar to the following:
+
+ ```go
+  2021-01-05T17:34:02.675Z|00004|acl_log(ovn_pinctrl0)|INFO|name="<Network Policy Name>", verdict=drop, severity=info: icmp,vlan_tci=0x0000,dl_src=50:54:00:00:00:02,dl_dst=50:54:00:00:00:01,nw_src=192.168.0.3,nw_dst=192.168.0.2,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0
+ ```
+
+#### Story 2
+
+As an Openshift user I want to control the rate of ACL logging to ensure I can extract the useful information involving accept/reject actions in high hit rate scenarios.
+
+#### Story 3
+
+As an Openshift user I want to monitor in realtime traffic that is hitting any network policies I have implemented.
+
+--- TBD in implementation (Realtime Monitoring would be nice here)
+
+### Implementation Details/Notes/Constraints
+
+The following codebase changes will need to done before this feature becomes usable:  
+
+- OVN Kubernetes
+
+  - Move from a cluster wide namespace default deny port group to namespace specific ones.
+  - Add logging flag and custom OVN meter to all created ACLs:
+
+   ```go
+    fmt.Sprintf("log=%v", aclLogging != ""),
+    fmt.Sprintf("severity=%s", getACLLoggingSeverity(aclLogging)), "meter=acl-logging",
+   ```
+
+  - Upstream work
+
+- Cluster Network Operator
+  
+  - Digest the `aclLoggingRateLimit` Flag and start OVN-K with the configured value.
+
+- Logging Stack
+  
+  - We need a way to collect and present ACL logging from the OVN controller logs.
```

Comment:
I think we can clarity this as a demo and documentation task.  I don't think any changes are required.  We just want to demonstrate the use case.

Reply:
ACK agreed